witness/README.md

```diff
@@ -1,15 +1,45 @@
-Witnessing
-============
+# Log Witnessing
 
-Witnessing is an important part of any transparent ecosystem, as it ensures that
-misbehavior on the part of logs can be prevented or detected. This is done by
-ensuring that the _checkpoints_ published by logs are _consistent_, which means 
-the log is presenting the same version of its contents to everyone and
-thus isn't carrying out a _split-view attack_.
+## Overview
 
-In terms of checkpoints, we consider witness implementations that use the
+Witnessing is an important part of log ecosystems that prevents logs from
+successfully accomplishing a _split-view attack_. A split-view attack is
+one in which the log prevents different populations with different views.
+Each of these views is append-only, and thus all clients checking consistency
+proofs are satisfied, yet the log is not fulfilling its claim of providing a
+_globally consistent_ list.
+
+For each witnessed log, a Witness will cosign log _checkpoints_ only if it
+can verify that it is consistent with the last cosigned checkpoint, i.e. that
+this new checkpoint represents only new values being appended to the log.
+Clients that acquire checkpoints signed by multiple independent witnesses can
+have more confidence that they are seeing a truly globally consistent log.
+
+## Roles
+
+We consider witnessing to be composed of three primary _roles_:
+  1. Feeder: acquires checkpoints and consistency proofs, and provides these
+     to the witness
+  2. Witness: maintains and cosigns a checkpoint for each log, and updates
+     this state to a new checkpoint only when convinced that the delta is only
+     appending new leaves
+  3. Distributor: makes cosigned checkpoints available to clients
+
+Note that this does not mean there needs to be 3 separate actors playing these roles.
+As a counterpoint, a common situation is to have the actor playing the witness role
+also polling the log for checkpoints and proofs (i.e. playing the feeder role), and
+making cosigned checkpoints available via some API (distributor).
+
+## Generic Witness Implementation
+
+This repository provides a witness that will work for any log that uses the
 [format defined in this
 repository](https://github.com/google/trillian-examples/tree/master/formats/log).
+Each log will require a custom feeder implementation to provide the binding
```

Comment:
It's probably less than "each log" right?
Perhaps more along the lines of "type of log" or something (although I'm not sure exactly what the right phrasing is)